# SingleCellTranscriptEvidence

## Introduction
The analysis of single cell RNA-Seq data involves the use of a disparate set of software packages to produce a list of genes that are expressed by the cell. Users can employ the pipeline described here to analyze sequences submitted to the Sequence Read Archive (SRA) to discover novel genes yet to be annotated. The workflow for the pipeline is in the figure shown below:

## Workflow, Use-case scenarios

## Dependencies
The pipeline expects the following packages are installed and available in the PATH:
* HISAT2 (version )
* StringTie2 (version )
* Python (version )
* ...

## Usage
main.py -acc <SRA_accession> -output

## Examples

## Output explanation

## Limitations/To-do

## People/Team
Ashis Saha <ashis@jhu.edu>
Michael Chambers <michael.chambers2@nih.gov>
Allissa Dillman <allissa.dillman@gmail.com>
Jessime Kirk <jessime@email.unc.edu>
Sara Kimiko Suzuki <sksuzuki@ad.unc.edu>
Wes Crouse <wcrouse@email.unc.edu>
Vamsi Kodali <vkkodali@gmail.com>
